src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala

```diff
@@ -28,8 +28,15 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
   val MAX_SHIFT: Int = 512
 
   lazy val wellDefExprAndType: WellDefinedness[PExpressionAndType] = createWellDef {
-
-    case _: PNamedOperand => noMessages // no checks to avoid cycles
+    case n: PNamedOperand =>
+      resolve(n) match {
+        /* A closure name can only be used as a spec, if we are not directly within the closure itself
+           (the same limitation applies within closures nested inside the closure itself) */
+        case Some(ap.Closure(id, _)) => error(n, s"here, the closure name ${n} can only be used as a spec",
+          !tree.parent(n).head.isInstanceOf[PClosureSpecInstance] &&
+            tryEnclosingFunctionLit(n).fold(true)(lit => lit.decl.id.fold(true)(encId => encId.name != id.name)))
```

Comment:
I do not think that you need such a strict restrictions. You could introduce an additional internal type, e.g. `spec` and then the usual type rules should enforce that a spec id is not used in a context that it should not be used in.

Reply:
I think this is a bit less restrictive than having a separate type `spec`.

A rule for a `spec` type, in my understanding, would be that operands can never be of type `spec`, and a `spec` cannot be the same as a `func` (tell me if that's wrong). But, within a closure declaration itself, it is useful to assert `c == name`, where `c` is some closure (either captured or argument), and `name` is the name of the declaration. This means that the verifier knows that the variables captured by `c` are the same as the `current` closure - this is necessary for recursion.

Initially, I also allowed the closure declaration name to be used outside of the declaration itself (in ghost code). But that turned out to be a lot of additional code in the desugarer, and not too useful - a user can always alias the literal by themselves, if they need to.

I have changed the message to `expected valid operand, got closure declaration name $n` to make it a bit more consistent with other error messages.